core/pkg/topology/topology.go

```diff
@@ -79,28 +88,96 @@ func (r *ring) Neighbours(token uint64) (prev *VNode, next *VNode) {
 	return
 }
 
-func (r *ring) VNodes() []*VNode {
+func (r *ring) Partitions(nodeId uint32) []Partition {
+	result := make([]Partition, 0)
+	r.vNodeDo(nodeId, func(i int, vNode VNode) {
+		for j := uint(1); j < r.replicationFactor; j++ {
+			i = r.prevIndex(i)
+		}
+
+		for j := uint(0); j < r.replicationFactor; j++ {
+			vNode := r.vNodes[i]
+			from := vNode.Token()
+			i = r.nextIndex(i)
+			partition := Partition{From: from, To: r.vNodes[i].Token() - 1, NodeId: vNode.NodeId()}
+			result = append(result, partition)
+		}
+	})
+
+	return result
+}
+
+func (r *ring) ExcessPartitions(nodeId uint32) []Partition {
+	partitions := r.Partitions(nodeId)
+
+	result := make([]Partition, 0)
+	for i := 0; i < len(partitions); i++ {
+		j := nextIndex(i, len(partitions))
+		if partitions[i].To == partitions[j].From-1 {
+			continue
+		}
+
+		result = append(result, Partition{From: partitions[i].To + 1, To: partitions[j].From - 1, NodeId: nodeId})
+	}
+
+	return result
+}
+
+func (r *ring) RemoveVNode(token uint64) bool {
+	vNodeId := r.search(token)
+	if vNodeId >= len(r.vNodes) || r.vNodes[vNodeId].Token() != token {
+		return false
+	}
+
+	copy(r.vNodes[vNodeId:], r.vNodes[vNodeId+1:])
+	r.vNodes[len(r.vNodes)-1] = VNode{}
```

Comment:
For what? 
r.vNodes = append(r.vNodes[: vNodeId], r.vNodes[r.vNodes+1:]...)

Reply:
updated